src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -395,90 +410,97 @@ PayChanClaim::doApply()
     Keylet const k (ltPAYCHAN, ctx_.tx[sfPayChannel]);
     auto const slep = ctx_.view ().peek (k);
     if (!slep)
-        return tecNO_TARGET;
+        return tecNO_ENTRY;
 
-    AccountID const src = (*slep)[sfAccount];
-    AccountID const dst = (*slep)[sfDestination];
-    AccountID const txAccount = ctx_.tx[sfAccount];
+    auto chanMembers = slep->getFieldArray (sfChannelMembers);
+    auto const& txClaims = ctx_.tx.getFieldArray (sfChannelClaims);
+    bool const thirdPartySubmission =
+        chanMembers[0].getAccountID (sfAccount) != ctx_.tx[sfAccount] &&
+        chanMembers[1].getAccountID (sfAccount) != ctx_.tx[sfAccount];
 
     auto const curExpiration = (*slep)[~sfExpiration];
+    if (curExpiration)
     {
-        auto const cancelAfter = (*slep)[~sfCancelAfter];
         auto const closeTime =
             ctx_.view ().info ().parentCloseTime.time_since_epoch ().count ();
-        if ((cancelAfter && closeTime >= *cancelAfter) ||
-            (curExpiration && closeTime >= *curExpiration))
+        if (closeTime >= *curExpiration)
             return closeChannel (
                 slep, ctx_.view (), k.key, ctx_.app.journal ("View"));
-    }
 
-    if (txAccount != src && txAccount != dst)
+        // PayChanClaim without a claim does nothing if
+        // the channel has a pending expiration
+        if (txClaims.empty ())
+            return tecNO_PERMISSION;
+    }
+    // third party must have claim if the channel has not expired
+    else if (thirdPartySubmission && txClaims.empty ())
         return tecNO_PERMISSION;
 
-    if (ctx_.tx[~sfBalance])
+    bool submittedOwnClaim = false;
+
+    if (! txClaims.empty ())
     {
-        auto const chanBalance = slep->getFieldAmount (sfBalance).xrp ();
-        auto const chanFunds = slep->getFieldAmount (sfAmount).xrp ();
-        auto const reqBalance = ctx_.tx[sfBalance].xrp ();
+        std::vector<STAmount> amounts {
+            chanMembers[0][sfAmount],
+            chanMembers[1][sfAmount]};
 
-        if (txAccount == dst && !ctx_.tx[~sfSignature])
-            return temBAD_SIGNATURE;
+        std::vector<std::uint32_t> sequences {
+            chanMembers[0][sfSequence],
+            chanMembers[1][sfSequence]};
 
-        if (ctx_.tx[~sfSignature])
+        for (auto const& txClaim : txClaims)
```

Comment:
I check that two claims are from different signers in `preflight`
https://github.com/wilsonianb/rippled/blob/bidir-pay-chan/src/ripple/app/tx/impl/PayChan.cpp#L376-L379

However, it looks like I need to add a check that there aren't more than two claims.

The loop can handle any order since it figures out which participant the claim belongs to here:
https://github.com/wilsonianb/rippled/blob/bidir-pay-chan/src/ripple/app/tx/impl/PayChan.cpp#L453-L457
As @nbougalis pointed out, the check should made clearer
https://github.com/ripple/rippled/pull/1869#discussion_r82893037